File: python/signalhouse/domains/groups.py
"""Groups domain for the SignalHouse SDK."""

from __future__ import annotations

from typing import Any, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from ..client import SignalHouseSDK


class GroupsAdmin:
    """Admin-only group operations (SignalHouse staff)."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk

    def get_groups(
        self,
        *,
        page: int | None = None,
        limit: int | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get a list of all groups with optional pagination.

        Args:
            page: The page number for pagination.
            limit: The number of items per page.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({"page": page, "limit": limit})
        return self._sdk._request(
            f"/group{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def create_group(
        self,
        group_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Create a new group with the specified group data.

        Args:
            group_data: The data for the new group, including required fields such as groupName.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If group_data is missing.
        """
        self._sdk._require({"groupData": group_data})
        return self._sdk._request(
            "/group",
            method="POST",
            body=group_data,
            token=token,
            headers=headers,
        )

    def delete_group(
        self,
        group_id: str,
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Delete a group with the specified group ID.

        Args:
            group_id: The ID of the group to delete.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If group_id is missing.
        """
        self._sdk._require({"groupId": group_id})
        safe_group_id = quote(str(group_id), safe="")
        return self._sdk._request(
            f"/group/{safe_group_id}",
            method="DELETE",
            token=token,
            headers=headers,
        )


class Groups:
    """Group management operations."""

    def __init__(self, sdk: SignalHouseSDK) -> None:
        self._sdk = sdk
        self.admin: GroupsAdmin | None = None
        if sdk.enable_admin:
            self.admin = GroupsAdmin(sdk)

    def get_group(
        self,
        *,
        id: str | None = None,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Get details of a group by its ID.

        Args:
            id: The ID of the group to retrieve.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.
        """
        query_string = self._sdk._get_query_string({"id": id})
        return self._sdk._request(
            f"/group{query_string}",
            method="GET",
            token=token,
            headers=headers,
        )

    def update_group(
        self,
        id: str,
        group_data: dict[str, Any],
        *,
        token: str | None = None,
        headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        """Update a group with the specified group data.

        Args:
            id: The ID of the group to update.
            group_data: The data for the group, including required fields such as groupName.
            token: Optional bearer token for authentication.
            headers: Additional headers to include in the request.

        Returns:
            Standardized response dict.

        Raises:
            SignalHouseValidationError: If id or group_data is missing.
        """
        self._sdk._require({"id": id, "groupData": group_data})
        safe_id = quote(str(id), safe="")
        return self._sdk._request(
            f"/group/{safe_id}",
            method="PUT",
            body=group_data,
            token=token,
            headers=headers,
        )

File: php/src/Domains/Groups.php
<?php

namespace SignalHouse\SDK\Domains;

use SignalHouse\SDK\HttpClient;

class Groups
{
    private HttpClient $client;
    private bool $enableAdmin;
    public ?object $admin = null;

    public function __construct(HttpClient $client, bool $enableAdmin)
    {
        $this->client = $client;
        $this->enableAdmin = $enableAdmin;

        if ($enableAdmin) {
            $this->admin = new class($client) {
                private HttpClient $client;

                public function __construct(HttpClient $client)
                {
                    $this->client = $client;
                }

                /**
                 * Get a list of all groups with optional pagination
                 */
                public function getGroups(array $params = [], array $options = []): array
                {
                    $queryString = $this->client->getQueryString($params);
                    return $this->client->request("/group{$queryString}", array_merge([
                        'method' => 'GET',
                    ], $options));
                }

                /**
                 * Create a new group
                 */
                public function createGroup(array $groupData, array $options = []): array
                {
                    $this->client->require(['groupData' => $groupData]);
                    return $this->client->request('/group', array_merge([
                        'method' => 'POST',
                        'body' => $groupData,
                    ], $options));
                }

                /**
                 * Delete a group
                 */
                public function deleteGroup(string $groupId, array $options = []): array
                {
                    $this->client->require(['groupId' => $groupId]);
                    $safeGroupId = rawurlencode($groupId);
                    return $this->client->request("/group/{$safeGroupId}", array_merge([
                        'method' => 'DELETE',
                    ], $options));
                }
            };
        }
    }

    /**
     * Get details of a group by its ID
     *
     * @param string $id The ID of the group
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function getGroup(string $id, array $options = []): array
    {
        $queryString = $this->client->getQueryString(['id' => $id]);
        return $this->client->request("/group{$queryString}", array_merge([
            'method' => 'GET',
        ], $options));
    }

    /**
     * Update a group with the specified data
     *
     * @param string $id The ID of the group
     * @param array $groupData The data to update
     * @param array $options Additional request options
     * @return array The response from the server
     */
    public function updateGroup(string $id, array $groupData, array $options = []): array
    {
        $this->client->require(['id' => $id, 'groupData' => $groupData]);
        $safeId = rawurlencode($id);
        return $this->client->request("/group/{$safeId}", array_merge([
            'method' => 'PUT',
            'body' => $groupData,
        ], $options));
    }
}
